### DISCUSSION TITLE
Why does primitive Box doesn't use special support-function but instead falls back to discrete vertices ?

### DISCUSSION THREAD

**Discussion by manuel-koch** [AlexanderFabisch/distance3d#74] [General]
I'm comparing distance3d performance with respect to using meshes ( vertices approximating a box, sphere, cylinder ) vs using primitives like `colliders.Box()`.
When doing many collision tests, I noticed that the former ( meshes ) appears to be faster than the latter ( primitives ).
Looking into `colliders.Box()` I noticed that it doesn't seem to use the generic support function `support_function_box` but instead falls back to plain vertices, whereas the primitive `colliders.Sphere` uses the builtin function `support_function_sphere`.

I.e. `colliders.ConvexHullVertices` colliding with `colliders.ConvexHullVertices` seems to be faster
than `colliders.Box|colliders.Sphere|colliders.Cylinder` colliding with `colliders.ConvexHullVertices`.

I even tried to patch `colliders.Box()`using the following implementation, but the results are the same ( `ConvexHullVertices`is faster than the other primitive types ):
```python
class Box(ConvexHullVertices):
   .....
   def support_function(self, search_direction):
        return support_function_box(search_direction, self.box2origin, self.size)
```

**Comment by AlexanderFabisch** [comment]
Box uses the vertex representation because it is faster AND correct. In the case of spheres, capsules, cylinders, etc. the vertex representation might be faster, but it is not entirely correct. We would need an infinitesimal resolution if we want to make it correct.

Having said that, I think the real support function should be faster if we had a properly compiled support function. Unfortunately, numba is not good at handling classes. Caching and inheritance are not supported. I am still interested in making the code faster, but at some point it is not possible anymore unfortunately...

**Comment by manuel-koch** [comment]
I'm currently testing other collision detection libs too and those that support primitives seem to perform better when using primitives instead of approximated mesh vertices:
- fcl
- pybullet

**Reply by AlexanderFabisch** [reply]
Of course, these are C++ implementations and the code can be much more efficient.

For this library I wanted to keep everything in pure Python to make it easily editable and to integrate in Python code without too many dependencies. We are currently exploring the limits of this approach with numba.

**Reply by AlexanderFabisch** [reply]
For comparison, standard FCL uses libccd as a backend. The ported libccd implementations (of GJK and MPR) are also part of this library.

**Reply by AlexanderFabisch** [reply]
pybullet internally also uses a fork of libccd.
